### PR TITLE
PHOENIX-5148

### PR DESCRIPTION
**Pull request by comnetwork**

### CODE REVIEW

**Comment by twdsilva** [comment]
@comnetwork  Thanks for the detailed explanation of your approach on the JIRA. I am reviewing the patch.

File: phoenix-core/src/main/java/org/apache/phoenix/compile/QueryPlan.java [hunk 1]
```diff
@@ -92,4 +94,17 @@
...
+     * Get the actual OrderBys of this queryPlan, which may be different from {@link #getOrderBy()},
+     * because {@link #getOrderBy()} is only the compiled result of {@link SelectStatement}.
+     * The return type is List because we can get multiple OrderBys for the query result of {@link SortMergeJoinPlan},
+     * eg. for the sql:
+     * SELECT  * FROM T1 JOIN T2 ON T1.a = T2.a and T1.b = T2.b
+     * The result of the sort-merge-join is sorted on (T1.a, T1.b) and (T2.a, T2.b) at the same time.
+     * </pre>
+     * @return
+     */
+    public List<OrderBy> getActualOutputOrderBys() ;
```
Comment (twdsilva) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/QueryPlan.java]: There would only be multiple OrderBys in the case of a join, right?

Comment (twdsilva) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/QueryPlan.java]: Also,  maybe we can come up with a better API name like ``getPhysicalOrderBy`` or ``getFinalOrderBy``, WDYT?

**Comment by twdsilva** [comment]
Overall looks pretty good. @dbwong or @BinShi-SecularBird would be good if you guys review this as well.

File: phoenix-core/src/main/java/org/apache/phoenix/compile/GroupByCompiler.java [hunk 2]
```diff
@@ -63,9 +63,10 @@
         private final boolean isOrderPreserving;
```
Comment (twdsilva) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/GroupByCompiler.java]: Do you need this boolean any more? Seems like we always should use ```orderPreservingTrackInfos``` to determine if its order preseving right?

File: phoenix-core/src/main/java/org/apache/phoenix/compile/OrderPreservingTracker.java [hunk 3]
```diff
@@ -9,29 +9,37 @@
...
 import org.apache.phoenix.expression.visitor.StatelessTraverseNoExpressionVisitor;
+import org.apache.phoenix.jdbc.PhoenixConnection;
 import org.apache.phoenix.schema.PTable;
-import org.apache.phoenix.schema.SortOrder;
+import org.apache.phoenix.schema.TableRef;
 import org.apache.phoenix.util.ExpressionUtil;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Iterators;
 import com.google.common.collect.Lists;
```
Comment (twdsilva) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/OrderPreservingTracker.java]: Can you explain the changes you had to make to this class, its a little hard to follow all the changes?

File: phoenix-core/src/main/java/org/apache/phoenix/execute/AggregatePlan.java [hunk 4]
```diff
@@ -327,4 +333,15 @@ public boolean useRoundRobinIterator() throws SQLException {
         return visitor.visit(this);
     }
 
+    private static OrderBy convertActualOutputOrderBy(OrderBy orderBy, GroupBy groupBy, StatementContext statementContext) {
```
Comment (twdsilva) [inline phoenix-core/src/main/java/org/apache/phoenix/execute/AggregatePlan.java]: Does every plan have to implement this function, with different logic?

File: phoenix-core/src/main/java/org/apache/phoenix/compile/QueryPlan.java (hunk 1 above)
Comment (comnetwork) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/QueryPlan.java]: Yes,now there would only be multiple OrderBys for SortMergeJoin.

Yes,I think the name `getFinalOrderBy` is OK , I named it as 'getActualOutputOrderBys' because
the equivalent name in Spark SQL's SparkPlan is `outputOrdering`, FYI.

File: phoenix-core/src/main/java/org/apache/phoenix/compile/GroupByCompiler.java (hunk 2 above)
Comment (comnetwork) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/GroupByCompiler.java]: Yes , I think the member variable `isOrderPreserving` is required because it saved the result of `OrderPreservingTracker.isOrderPreserving()`.  `GroupBy.isOrderPreserving()` is used by many classes under compile and execute package.
FYI, the `GroupBy` class is a immutable class , when we compile the `group by clause` at difference time(we compile it twice in `QueryCompiler.compileSingleFlatQuery`) , a new GroupBy instance  is created.

File: phoenix-core/src/main/java/org/apache/phoenix/compile/OrderPreservingTracker.java (hunk 3 above)
Comment (comnetwork) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/OrderPreservingTracker.java]: 1. I added a getInputOrderBys method to determine the input OrderBys by the combination of innerQueryPlan's output OrderBys ， GroupBy of current QueryPlan or the rowKeyColumns of current table.

2. Because the innerQueryPlan may have multiple output OrderBys(for `SortMergeJoinPlan`), so I extracted many stateful memeber variables in `OrderPreservingTracker`(such as `orderPreservingTrackInfos`, `isOrderPreserving` and `isReverse` etc.) to a new inner class `TrackOrderByContext`, `TrackOrderByContext` is used to track if a single Input OrderBy matchs
the target OrderByExpressions in `OrderPreservingTracker.isOrderPreserving()`, once we found a `TrackOrderByContext`  satisfied `isOrderPreserving()`,
`OrderPreservingTracker.selectedTrackOrderByContext` is set to this `TrackOrderByContext`, then we can use this `TrackOrderByContext` to implement the `OrderPreservingTracker.getOrderPreservingTrackInfos` and `OrderPreservingTracker. isReverse` etc.

File: phoenix-core/src/main/java/org/apache/phoenix/execute/AggregatePlan.java (hunk 4 above)
Comment (comnetwork) [reply phoenix-core/src/main/java/org/apache/phoenix/execute/AggregatePlan.java]: Not every plan have to  implement this function, eg. HashJoinPlan and SortMergeJoinPlan.
In my opinion, this method is not suitable to declared as abstract method in QueryPlan and subclass to override it , because this method is to invoke in the ctor of concrete  subclass.

File: phoenix-core/src/main/java/org/apache/phoenix/compile/QueryPlan.java (hunk 1 above)
Comment (twdsilva) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/QueryPlan.java]: ``getOutputOrdering`` also works.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/SortMergeJoinMoreIT.java [hunk 5]
```diff
@@ -423,7 +424,7 @@ public void testBug2894() throws Exception {
                         "         ) L\n" +
                         "     ON L.BUCKET = E.BUCKET AND L.TIMESTAMP = E.TIMESTAMP\n" +
                         " ) C\n" +
-                        " GROUP BY C.BUCKET, C.TIMESTAMP";
+                        " GROUP BY C.BUCKET, C.TIMESTAMP ORDER BY C.BUCKET, C.TIMESTAMP";
```
Comment (twdsilva) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/SortMergeJoinMoreIT.java]: Why do we need to add the order by here?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/join/SubqueryUsingSortMergeJoinIT.java [hunk 6]
```diff
@@ -516,8 +516,8 @@ public void testAnyAllComparisonSubquery() throws Exception {
             assertEquals(rs.getString(2), "T6");
 
             assertFalse(rs.next());
-            
-            query = "SELECT /*+ USE_SORT_MERGE_JOIN*/ \"order_id\", name FROM " + tableName4 + " o JOIN " + tableName1 + " i ON o.\"item_id\" = i.\"item_id\" WHERE quantity != ANY(SELECT quantity FROM " + tableName4 + " q WHERE o.\"item_id\" = q.\"item_id\" GROUP BY quantity)";
+            //add order by to make the query result stable
+            query = "SELECT /*+ USE_SORT_MERGE_JOIN*/ \"order_id\", name FROM " + tableName4 + " o JOIN " + tableName1 + " i ON o.\"item_id\" = i.\"item_id\" WHERE quantity != ANY(SELECT quantity FROM " + tableName4 + " q WHERE o.\"item_id\" = q.\"item_id\" GROUP BY quantity) order by \"order_id\"";
```
Comment (twdsilva) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/join/SubqueryUsingSortMergeJoinIT.java]: what happens if we don't add the order by?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/SortMergeJoinMoreIT.java (hunk 5 above)
Comment (comnetwork) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/SortMergeJoinMoreIT.java]: I added the order by here mainly for:
1.Make the query result stable,because the original sql is only have the group by. The group by is not `orderPreserving` without this patch and must add a additional `SORTED BY [E.TIMESTAMP, E.BUCKET]`, but after this patch, the group by is  `orderPreserving`, so the query result is not same as without this patch.
2. To test if the GroupBy and OrderBy could both be optimized, from the explain we can see the 
`GROUP BY C.BUCKET, C.TIMESTAMP` is  `orderPreserving` and `ORDER BY C.BUCKET, C.TIMESTAMP` is compiled out.

If I do not add the order by, the explain assertion and query result assertion is same as the order by is added. 

Another curious problem  is that  why the order of query result is reversed after this patch:
the query result without this patch is :
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993520000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993515000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993510000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993505000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993490000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993485000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993480000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993475000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993470000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993430000000000L, rs.getLong(2));
the query result after this patch is :
               assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993430000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993470000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993475000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993480000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993485000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993490000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993505000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993510000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993515000000000L, rs.getLong(2));
                assertTrue(rs.next());
                assertEquals("5SEC", rs.getString(1));
                assertEquals(1462993520000000000L, rs.getLong(2));

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/join/SubqueryUsingSortMergeJoinIT.java (hunk 6 above)
Comment (comnetwork) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/join/SubqueryUsingSortMergeJoinIT.java]: If we don't add the order by , the order of the query result is reverse after this patch, just as:

File: phoenix-core/src/main/java/org/apache/phoenix/compile/QueryPlan.java (hunk 1 above)
Comment (comnetwork) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/QueryPlan.java]: @twdsilva , thank you for your opinion, in order to in correspondence with the existing name `getOrderBy()`  in `QueryPlan` class, can we named it as `getOutputOrderBy` ?

Comment (twdsilva) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/QueryPlan.java]: Sure that works

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/SortMergeJoinMoreIT.java (hunk 5 above)
Comment (twdsilva) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/SortMergeJoinMoreIT.java]: Ok, I see now the results are the same even if you don't add the order by and the order by gets optimized out.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/join/SubqueryUsingSortMergeJoinIT.java (hunk 6 above)
Comment (twdsilva) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/join/SubqueryUsingSortMergeJoinIT.java]: IMO its fine to not have the ```ORDER BY``` and change the test to verify that the order of results is reversed as previously the ```CLIENT SORTED BY [O.ITEM_ID]``` was not needed.

Comment (comnetwork) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/join/SubqueryUsingSortMergeJoinIT.java]: @twdsilva , thank you for your opinion, the reason why the query result with this patch is not same as before this patch is different from previous test, if there are more than two tuples which O.ITEM_ID are '0000000006', the query result may not be accurately reversed. Following your suggestion, I would add explain assertions to verify the unnecessary sort is avoid.

**Comment by comnetwork** [comment]
@twdsilva , I made a new PR #468  for this JIRA, please help me have a review again, thank you very much.

**Comment by gjacoby126** [comment]
JIRA resolved, closing PR
